### PR TITLE
Issue 94 - Now with Presence Validators for All Environment Variables

### LINKED ISSUE #94: Add better error messages for setup in general
[Just generally more informative messages in general. You can start from lib/github-integration/install.js.]

1. If any necessary env vars are missing, let the dev know which ones are missing and what the developer needs to do
1. There was something else, but I don't remember what

### PR DESCRIPTION
**Pull request by rpigneri-vol**
I have added code to validate that all environment variables are present and present readable error messages to the user.   This code does not test the correctness of the values of the environment variables.

Close #94.

### CODE REVIEW

File: lib/utils/env_vars.js [hunk 1]
```diff
@@ -24,6 +32,50 @@ const env_vars = {
...
+/** 
+ * For each field in env_vars, check if the value is defined in an environment
+ * variable, and throw an exception if it is not.
+ * 
+ * @throws ReferenceError with a message for all missing variables. 
+ */
+ env_vars.validateEnvironment = function() {
+
+  // We are checking for BASE_URL, but we are referring it as SERVER_URL as that
+  // is the name used in GitHub Actions.
```
Comment (plocket) [inline lib/utils/env_vars.js:44]:
```suggestion
// is the name used in GitHub secrets, which is what most devs will be using.
```

File: lib/utils/env_vars.js [hunk 2]
```diff
@@ -24,6 +32,50 @@ const env_vars = {
...
+  {
+    errorMessage += '\nThe PLAYGROUND_EMAIL GitHub secret must be defined.  The PLAYGROUND_EMAIL is the e-mail address of the test user to run the Docassemble suite.'
+  }
+  if (!process.env.PLAYGROUND_PASSWORD)
+  {
+    errorMessage += '\nThe PLAYGROUND_PASSWORD GitHub secret must be defined.  The PLAYGROUND_PASSWORD is your test user\'s password for the test suite.'
+  }
+  if (!process.env.BASE_URL)
+  {
+    errorMessage += 'The SERVER_URL GitHub secret must be defined.  the SERVER_URL is the URL of the Docassemble Server where the tests should be run.';
```
Comment (plocket) [inline lib/utils/env_vars.js:64]:
```suggestion
    errorMessage += `\nThe SERVER_URL GitHub secret must be defined. the SERVER_URL is the URL of the Docassemble Server where the tests should be run.`;
```

File: lib/utils/env_vars.js [hunk 3]
```diff
@@ -24,6 +32,50 @@ const env_vars = {
...
+  }
+  if (!process.env.REPO_URL)
+  {
+    errorMessage += "The REPO_URL GitHub secret must be defined.  The REPO_URL is the URL to the Git repo whose Actions are being run.";
```
Comment (plocket) [inline lib/utils/env_vars.js:68]:
```suggestion
    errorMessage += `\nThe REPO_URL GitHub secret must be defined. The REPO_URL is the URL to the Git repo whose Actions are being run.`;
```

File: tests/unit_tests/env_vars.test.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+const expect = chai.expect;
+
+const env_vars = require("../../lib/utils/env_vars.js");
+const validateEnvironment = env_vars.validateEnvironment;
+
+describe('env_vars', function () {
+    describe('validateEnvironment', function() {
+
+        /**
+         * All variables should be defined in .env.  For reference, those variables are the following:
```
Comment (plocket) [inline tests/unit_tests/env_vars.test.js:11]:
```suggestion
* All required variables should be defined in .env. For reference, those variables are the following:
```

File: tests/unit_tests/env_vars.test.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            delete process.env.PLAYGROUND_ID;
+            delete process.env.PLAYGROUND_EMAIL;
+            delete process.env.PLAYGROUND_PASSWORD;
+            delete process.env.BASE_URL;
+            delete process.env.REPO_URL;
+
+            // This test just tests that the variable names show up in a fixed order.
+            //
+            // For whatever reason, /PLAYGROUND_ID.*PLAYGROUND_EMAIL.*PLAYGROUND_PASSWORD.*SERVER_URL.*REPO_URL/ simply wouldn't work once I got to the first letter
+            // after the second underscore so I have used separate matchers.
```
Comment (plocket) [inline tests/unit_tests/env_vars.test.js:45]: I think this doesn't quite do that, so let's make an issue to remove these comments.

File: tests/unit_tests/env_vars.test.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
+const chai = require('chai');
+const expect = chai.expect;
+
+const env_vars = require("../../lib/utils/env_vars.js");
+const validateEnvironment = env_vars.validateEnvironment;
```
Comment (plocket) [inline tests/unit_tests/env_vars.test.js:5]: TODO: Make an issue to mock the `process.env` vars.

Comment (plocket) [reply tests/unit_tests/env_vars.test.js:5]: Or just set the values.

Comment (plocket) [reply tests/unit_tests/env_vars.test.js:5]: So that tests don't depend on previous values.

**Review by plocket — APPROVED** [review]
Approved once suggestions are resolved and issues have been made.

File: README.md [hunk 7]
```diff
@@ -1,2 +1,8 @@
 # docassemble-cucumber
 Integrated automated end-to-end testing with docassemble, puppeteer, and cucumber.
+
+To setup the environment:
+> npm run setup
+
+To run the unit tests in isolation:
+> npm run unit
```
Comment (plocket) [inline README.md:8]: TODO: Make an issue to provide full testing instructions

File: tests/unit_tests/env_vars.test.js (hunk 6 above)
Comment (rpigneri-vol) [reply tests/unit_tests/env_vars.test.js:5]: Here is new [Issue 377](https://github.com/plocket/docassemble-cucumber/issues/377) to cover this area of concern.

**Comment by plocket** [comment]
Closes #363

**Pull request merged**